tokio/src/net/util/io_resource.rs

```diff
@@ -0,0 +1,380 @@
+/// While violating this requirement is "safe" from a Rust memory model point
+/// of view, it will result in unexpected behavior in the form of lost
+/// notifications and tasks hanging.
+///
+/// ## Readiness events
+///
+/// Besides just providing [`AsyncRead`] and [`AsyncWrite`] implementations,
+/// this type also supports access to the underlying readiness event stream.
+/// While similar in function to what [`Registration`] provides, the semantics
+/// are a bit different.
+///
+/// Two functions are provided to access the readiness events:
+/// [`poll_read_ready`] and [`poll_write_ready`]. These functions return the
+/// current readiness state of the `IoResource` instance. If
+/// [`poll_read_ready`] indicates read readiness, immediately calling
+/// [`poll_read_ready`] again will also indicate read readiness.
+///
+/// When the operation is attempted and is unable to succeed due to the I/O
+/// resource not being ready, the caller must call [`clear_read_ready`] or
+/// [`clear_write_ready`]. This clears the readiness state until a new
+/// readiness event is received.
+///
+/// This allows the caller to implement additional functions. For example,
+/// [`TcpListener`] implements poll_accept by using [`poll_read_ready`] and
+/// [`clear_read_ready`].
+///
+/// TODO
+/// ```rust
+/// use tokio::net::util::IoResource;
+///
+/// use futures_core::ready;
+/// use tokio::net::driver::Readiness;
+/// use mio::net::{TcpStream, TcpListener};
+/// use std::io;
+/// use std::task::{Context, Poll};
+///
+/// struct MyListener {
+///     io_resource: IoResource<TcpListener>,
+/// }
+///
+/// impl MyListener {
+///     pub fn poll_accept(&mut self, cx: &mut Context<'_>) -> Poll<Result<TcpStream, io::Error>> {
+///         let readiness = Readiness::READABLE;
+///
+///         ready!(self.io_resource.poll_read_ready(cx))?;
+///
+///         match self.io_resource.get_ref().accept() {
+///             Ok((socket, _)) => Poll::Ready(Ok(socket)),
+///             Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+///                 self.io_resource.clear_read_ready(cx)?;
+///                 Poll::Pending
+///             }
+///             Err(e) => Poll::Ready(Err(e)),
+///         }
+///     }
+/// }
+/// ```
+///
+/// TODO
+/// ## Platform-specific events?
+///
+/// [`std::io::Read`]: https://doc.rust-lang.org/std/io/trait.Read.html
+/// [`std::io::Write`]: https://doc.rust-lang.org/std/io/trait.Write.html
+/// [`AsyncRead`]: ../io/trait.AsyncRead.html
+/// [`AsyncWrite`]: ../io/trait.AsyncWrite.html
+/// [`mio::event::Source`]: TODO
+/// [`Registration`]: struct.Registration.html
+/// [`TcpListener`]: ../net/struct.TcpListener.html
+/// [`clear_read_ready`]: #method.clear_read_ready
+/// [`clear_write_ready`]: #method.clear_write_ready
+/// [`poll_read_ready`]: #method.poll_read_ready
+/// [`poll_write_ready`]: #method.poll_write_ready
+pub struct IoResource<S: mio::event::Source> {
+    io: Option<S>,
```

Comment:
Since you are in here, can `Option` be removed?

Reply:
It can! I was going to do it as part of this PR, but since `IoResource::into_inner` moves the value out, the `Drop` impl requires some additional handling to forget the value when this occurs. I can include it now, but had it as a note for a follow-up. It should be a fairly small change.